File: apps/backend/src/api/routes/preferences.py
from fastapi import APIRouter, Depends

from api.dependencies import get_preferences_service
from models.preferences import PreferencesResponse, UpdatePreferencesRequest
from services.preferences_service import PreferencesService

SETTINGS_PREFIX = "/preferences"

router = APIRouter(prefix=SETTINGS_PREFIX, tags=["preferences"])


@router.get("", response_model=PreferencesResponse)
def get_preferences(preferences_service: PreferencesService = Depends(get_preferences_service)):
    return preferences_service.get_preferences()


@router.patch("", response_model=PreferencesResponse)
def update_preferences(
        request: UpdatePreferencesRequest,
        preferences_service: PreferencesService = Depends(get_preferences_service),
):
    return preferences_service.update_preferences(request)

File: apps/frontend/api/preferences.ts
import client from "@/api/client";

const api = client.extend((options) =>
    ({prefix: `${options.prefix}/preferences`})
);

export type Theme = "light" | "dark" | "system";
export type Language = "en" | "de";

export interface Preferences {
    theme: Theme;
    language: Language;
    notificationsEnabled: boolean;
}

export interface UpdatePreferencesRequest {
    theme?: Theme | null;
    language?: Language | null;
    notificationsEnabled?: boolean | null;
}

export const preferencesApi = {
    getPreferences: async (): Promise<Preferences> => {
        return await api.get("").json();
    },

    updatePreferences: async (request: UpdatePreferencesRequest): Promise<Preferences> => {
        return await api.patch("", {json: request}).json();
    },
};
